File: server/PoultryFarmManager.Core/Models/SaleOrder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using PoultryFarmManager.Core.Enums;
using PoultryFarmManager.Core.Models.Finance;

namespace PoultryFarmManager.Core.Models;

public class SaleOrder : DbEntity
{
    public Guid BatchId { get; set; }
    public Batch? Batch { get; set; }

    public Guid CustomerId { get; set; }
    public Person? Customer { get; set; }

    public DateTime Date { get; set; }
    public SaleOrderStatus Status { get; set; } = SaleOrderStatus.Pending;
    public decimal PricePerKg { get; set; }
    public string? Notes { get; set; }

    public ICollection<SaleOrderItem> Items { get; set; } = [];
    public ICollection<Transaction> Payments { get; set; } = [];

    // Computed properties
    public decimal TotalAmount => Items.Sum(i => i.Weight) * PricePerKg;
    public decimal TotalPaid => Payments.Where(p => p.Type == TransactionType.Income).Sum(p => p.TransactionAmount);
    public decimal PendingAmount => TotalAmount - TotalPaid;
}

File: client/webapp/src/types/saleOrder.ts
export type SaleOrderStatus = "Pending" | "PartiallyPaid" | "Paid" | "Cancelled";

export interface SaleOrderItem {
    id: string;
    weight: number;
    unitOfMeasure: string;
    processedDate: string;
}

export interface SaleOrderPayment {
    transactionId: string;
    date: string;
    amount: number;
    notes: string | null;
}

export interface SaleOrder {
    id: string;
    batchId: string;
    batchName: string | null;
    customerId: string;
    customerFullName: string;
    date: string;
    status: SaleOrderStatus;
    notes: string | null;
    pricePerUnit: number;
    items: SaleOrderItem[];
    payments: SaleOrderPayment[];
    totalWeight: number;
    totalAmount: number;
    totalPaid: number;
    pendingAmount: number;
}

export interface NewSaleOrderItem {
    weight: number;
    unitOfMeasure: string;
    processedDateClientIsoString: string;
}

export interface NewSaleOrder {
    batchId: string;
    customerId: string;
    dateClientIsoString: string;
    pricePerUnit: number;
    items: NewSaleOrderItem[];
    notes: string | null;
}

export interface NewSaleOrderPayment {
    dateClientIsoString: string;
    amount: number;
    notes: string | null;
}
